codec-http/src/test/java/io/netty/handler/codec/http/websocketx/extensions/compression/PerMessageDeflateDecoderTest.java

```diff
@@ -88,51 +90,59 @@ public void testNormalFrame() {
 
         BinaryWebSocketFrame frame = new BinaryWebSocketFrame(true,
-                WebSocketExtension.RSV3, Unpooled.wrappedBuffer(payload));
+                WebSocketExtension.RSV3, decoderChannel.bufferAllocator().copyOf(payload));
 
         // execute
         assertTrue(decoderChannel.writeInbound(frame));
         BinaryWebSocketFrame newFrame = decoderChannel.readInbound();
 
         // test
         assertNotNull(newFrame);
-        assertNotNull(newFrame.content());
+        assertNotNull(newFrame.binaryData());
         assertEquals(WebSocketExtension.RSV3, newFrame.rsv());
-        assertEquals(300, newFrame.content().readableBytes());
+        assertEquals(300, newFrame.binaryData().readableBytes());
 
         byte[] finalPayload = new byte[300];
-        newFrame.content().readBytes(finalPayload);
+        readBytes(newFrame.binaryData(), finalPayload);
         assertArrayEquals(finalPayload, payload);
-        newFrame.release();
+        newFrame.close();
     }
 
     @Test
     public void testFragmentedFrame() {
-        EmbeddedChannel encoderChannel = new EmbeddedChannel(
-                ZlibCodecFactory.newZlibEncoder(ZlibWrapper.NONE, 9, 15, 8));
+        EmbeddedChannel encoderChannel = new EmbeddedChannel(BYTEBUF_TO_BUFFER_HANDLER,
+                ZlibCodecFactory.newZlibEncoder(ZlibWrapper.NONE, 9, 15, 8),
+                BUFFER_TO_BYTEBUF_HANDLER);
         EmbeddedChannel decoderChannel = new EmbeddedChannel(new PerMessageDeflateDecoder(false));
 
         // initialize
         byte[] payload = new byte[300];
         random.nextBytes(payload);
 
-        assertTrue(encoderChannel.writeOutbound(Unpooled.wrappedBuffer(payload)));
-        ByteBuf compressedPayload = encoderChannel.readOutbound();
-        compressedPayload = compressedPayload.slice(0, compressedPayload.readableBytes() - 4);
-
-        int oneThird = compressedPayload.readableBytes() / 3;
-        BinaryWebSocketFrame compressedFrame1 = new BinaryWebSocketFrame(false,
-                WebSocketExtension.RSV1 | WebSocketExtension.RSV3,
-                compressedPayload.slice(0, oneThird));
-        ContinuationWebSocketFrame compressedFrame2 = new ContinuationWebSocketFrame(false,
-                WebSocketExtension.RSV3, compressedPayload.slice(oneThird, oneThird));
-        ContinuationWebSocketFrame compressedFrame3 = new ContinuationWebSocketFrame(true,
-                WebSocketExtension.RSV3, compressedPayload.slice(oneThird * 2,
-                        compressedPayload.readableBytes() - oneThird * 2));
+        assertTrue(encoderChannel.writeOutbound(encoderChannel.bufferAllocator().copyOf(payload)));
+        Buffer compressedPayload = encoderChannel.readOutbound();
+        BinaryWebSocketFrame compressedFrame1;
+        ContinuationWebSocketFrame compressedFrame2;
+        ContinuationWebSocketFrame compressedFrame3;
+        try {
+            compressedPayload = compressedPayload.copy(0, compressedPayload.readableBytes() - 4);
```

Comment:
We forget to close the original `compressedPayload` when we overwrite it like that.

Reply:
Maybe you want to call `skipWritable(-4)` instead, in these tests.